### PR TITLE
[wip] Add basic mypy support

### PR DESCRIPTION
**Pull request by toabctl**
First step for mypy. Still type hints are missing for nearly
everything.

Signed-off-by: Thomas Bechtold <tbechtold@suse.com>

### CODE REVIEW

File: teuthology/misc.py [hunk 1]
```diff
@@ -1057,7 +1057,7 @@ def get_valgrind_args(testdir, name, preamble, v):
         return preamble
     if not isinstance(v, list):
         v = [v]
-    val_path = '/var/log/ceph/valgrind'.format(tdir=testdir)
+    val_path = '/var/log/ceph/valgrind'
```
Comment (kshtsk) [inline teuthology/misc.py]: Is this related to the title of the PR? It looks for me it is not. If so, should be in separate commit.

File: teuthology/orchestra/console.py [hunk 2]
```diff
@@ -266,8 +266,7 @@ def power_off_for_interval(self, interval=30):
         child = self._pexpect_spawn_ipmi('power on')
         child.expect('Chassis Power Control: Up/On', timeout=self.timeout)
         self._wait_for_login()
-        log.info('Power off for {i} seconds completed'.format(
-            s=self.shortname, i=interval))
+        log.info('Power off for {i} seconds completed'.format(i=interval))
```
Comment (kshtsk) [inline teuthology/orchestra/console.py]: Same as above, should go to cleanup commit.

File: teuthology/orchestra/console.py [hunk 3]
```diff
@@ -377,5 +376,4 @@ def power_off_for_interval(self, interval=30):
         self.vm_domain.info().destroy()
         time.sleep(interval)
         self.vm_domain.info().create()
-        log.info('Power off for {i} seconds completed'.format(
-            s=self.shortname, i=interval))
+        log.info('Power off for {i} seconds completed'.format(i=interval))
```
Comment (kshtsk) [inline teuthology/orchestra/console.py]: Same as above.

File: teuthology/orchestra/console.py (hunk 2 above)
Comment (kshtsk) [reply teuthology/orchestra/console.py]: I'm not sure maybe we need to leave 's', and add back it to the message?
Btw, since it is log, there can be better to use "%s" notation, like `log.info('some %s and %s', x, y)`

File: teuthology/provision/cloud/base.py [hunk 4]
```diff
@@ -13,7 +13,7 @@
 
 
 class Provider(object):
-    _driver_posargs = list()
+    _driver_posargs = list() # type: ignore
```
Comment (kshtsk) [inline teuthology/provision/cloud/base.py]: How does it work, why is this required, I am actually not familiar with 'mypy'.

File: teuthology/test/task/__init__.py [hunk 5]
```diff
@@ -1,5 +1,9 @@
 from mock import patch, DEFAULT
 from pytest import raises
+try:
+    from typing import Any
```
Comment (kshtsk) [inline teuthology/test/task/__init__.py]: why is this required?

File: teuthology/misc.py (hunk 1 above)
Comment (toabctl) [reply teuthology/misc.py]: It is. Without that change, `mypy` will complain that this code is not correct.

File: teuthology/orchestra/console.py (hunk 2 above)
Comment (toabctl) [reply teuthology/orchestra/console.py]: same here. the current code is incorrect and `mypy` complains about it

File: teuthology/orchestra/console.py (hunk 3 above)
Comment (toabctl) [reply teuthology/orchestra/console.py]: same answer as above

File: teuthology/test/task/__init__.py (hunk 5 above)
Comment (toabctl) [reply teuthology/test/task/__init__.py]: It's required for the type hints (even if these are only comments for py2).

**Comment by toabctl** [comment]
`mypy` is a static type checker. So if you add type hints to your classes/methods/functions/variables, `mypy` can check the code (without executing it) if the variables/functions/.. are used correctly.
See http://mypy-lang.org/ and https://docs.python.org/3/library/typing.html
Given that we have nearly no type hints yet in the code, it's not that useful (but also does not harm). But when adding new code, we can also add the type hints directly to get better test coverage.
btw. this is already used in ceph/ceph (eg. ceph-common and ceph-daemon) so it's nothing new to the ceph ecosystem.

File: teuthology/provision/cloud/base.py (hunk 4 above)
Comment (toabctl) [reply teuthology/provision/cloud/base.py]: See answer below.

**Comment by toabctl** [comment]
@kshtsk it's not WIP imo. I think it's ready to get merged and in follow up commits, the type hints can be added.
